modules/tinymce/src/core/main/ts/api/html/DomParser.ts

```diff
@@ -406,137 +406,10 @@ const DomParser = (settings?: DomParserSettings, schema = Schema()): DomParser =
-        if (schema.isValidChild(parent.name, node.name)) {
-          // Reverse the array since it makes looping easier
-          parents.reverse();
-
-          // Clone the related parent and insert that after the moved node
-          newParent = filterNode(parents[0].clone(), nodeFilters, attributeFilters, state);
-
-          // Start cloning and moving children on the left side of the target node
-          let currentNode = newParent;
-          for (let i = 0; i < parents.length - 1; i++) {
-            if (schema.isValidChild(currentNode.name, parents[i].name)) {
-              tempNode = filterNode(parents[i].clone(), nodeFilters, attributeFilters, state);
-              currentNode.append(tempNode);
-            } else {
-              tempNode = currentNode;
-            }
-
-            for (let childNode = parents[i].firstChild; childNode && childNode !== parents[i + 1];) {
-              const nextNode = childNode.next;
-              tempNode.append(childNode);
-              childNode = nextNode;
-            }
-
-            currentNode = tempNode;
-          }
-
-          if (!isEmpty(schema, nonEmptyElements, whitespaceElements, newParent)) {
-            parent.insert(newParent, parents[0], true);
-            parent.insert(node, newParent);
-          } else {
-            parent.insert(node, parents[0], true);
-          }
-
-          // Check if the element is empty by looking through it's contents and special treatment for <p><br /></p>
-          parent = parents[0];
-          if (isEmpty(schema, nonEmptyElements, whitespaceElements, parent) || hasOnlyChild(parent, 'br')) {
-            parent.empty().remove();
-          }
-        } else {
-          removeOrUnwrapInvalidNode(node);
-        }
-      } else if (node.parent) {
-        // If it's an LI try to find a UL/OL for it or wrap it
-        if (node.name === 'li') {
-          let sibling = node.prev;
-          if (sibling && (sibling.name === 'ul' || sibling.name === 'ol')) {
-            sibling.append(node);
-            continue;
-          }
-
-          sibling = node.next;
-          if (sibling && (sibling.name === 'ul' || sibling.name === 'ol')) {
-            sibling.insert(node, sibling.firstChild, true);
-            continue;
-          }
-
-          node.wrap(filterNode(new AstNode('ul', 1), nodeFilters, attributeFilters, state));
-          continue;
-        }
-
-        // Try wrapping the element in a DIV
-        if (schema.isValidChild(node.parent.name, 'div') && schema.isValidChild('div', node.name)) {
-          node.wrap(filterNode(new AstNode('div', 1), nodeFilters, attributeFilters, state));
-        } else {
-          // We failed wrapping it, remove or unwrap it
-          removeOrUnwrapInvalidNode(node);
-        }
-      }
+      extractInvalidNode(nodes[ni], schema, fixedNodes, (newNode) => filterNode(newNode, nodeFilters, attributeFilters, state));
```

Comment:
This adds potentially a lot of context/stack switches which could be a performance issue here... can we change this to pass in an array instead so we avoid some of this?

Reply:
Done — also helped me simplify the arguments to the function a little bit, thankfully